torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -640,6 +681,172 @@ def _need_prefetch_post_backward_hook(self) -> bool:
+                        "All FSDP module should the same state_dict_type."
+                    )
+                module.state_dict_type = state_dict_type
+        try:
+            yield
+        finally:
+            for module in self.modules():
+                if isinstance(module, FullyShardedDataParallel):
+                    module.state_dict_type = prev_state_dict_type
+
+    @staticmethod
+    def _post_state_dict_hook(
+        module: nn.Module,
+        state_dict: "OrderedDict[str, torch.Tensor]",
+        prefix: str,
+        *args: Any,
+    ) -> "OrderedDict[str, torch.Tensor]":
+        """
+        _post_state_dict_hook() is called after the state_dict() of this
+        FSDP module is executed. ``self.state_dict_type`` is used to decide
+        what postprocessing will be done.
+        """
+        self = cast(FullyShardedDataParallel, module)
+        if self.state_dict_type == StateDictType.STATE_DICT:
+            raise NotImplementedError("Will be implemented in the next PRs.")
+        elif self.state_dict_type == StateDictType.LOCAL_STATE_DICT:
+            flat_param = getattr(self.module, "flat_param", None)
+            full_numel = flat_param.full_numel
+            assert (
+                flat_param is not None
+            ), "flat_param cannot be None when doing local_state_dict."
+            shard_offset = flat_param.numel() * self.rank
+            valid_data_size = flat_param.numel() - flat_param.num_padded
+            if valid_data_size > 0 and flat_param.num_padded > 0:
+                flat_param = flat_param.narrow(0, 0, valid_data_size)
+            local_shards = [
+                Shard.from_tensor_and_offsets(flat_param, [shard_offset], self.rank)
+            ]
+            state_dict[f"{prefix}flat_param"] = init_from_local_shards(
+                local_shards, full_numel, process_group=self.process_group
+            )  # type: ignore[assignment]
+        elif self.state_dict_type == StateDictType.SHARDED_STATE_DICT:
+            raise NotImplementedError("Will be implemented in the next PRs.")
+        else:
+            raise ValueError(f"Unknown StateDictType {self.state_dict_type}.")
+        return state_dict
+
+    def state_dict(self, destination=None, prefix="", keep_vars=False):
+        """
+        The entry point of all three FSDP state_dict APIs.
+        ``self.state_dict_type`` decides which code path to execute.
+
+        .. warning:: This needs to be called on all ranks, since synchronization
+            primitives may be used.
+        """
+        torch.cuda.synchronize()
+        if self.state_dict_type == StateDictType.STATE_DICT:
+            raise NotImplementedError("Will be implemented in the next PRs.")
+        elif self.state_dict_type == StateDictType.LOCAL_STATE_DICT:
+            state_dict = self.module.flat_state_dict(destination, prefix, keep_vars)
+            state_dict = self._post_state_dict_hook(self, state_dict, prefix=prefix)
```

Comment:
Why do we need to manually call this if it is registered as a hook in FSDP ctor?

Reply:
Fixed the inconsistency.